src/many-modules/src/_9_account/features/multisig.rs

```diff
@@ -9,112 +9,189 @@ use many_macros::many_module;
+            MemoInner::ByteString(bstr) => Some(bstr.as_slice()),
+        })
     }
 }
 
-impl Memo<String> {
-    // USED ONLY IN TESTS! NEVER REMOVE THIS GUARD! MEMOS SHOULDN'T BE PRODUCED FROM LARGE STRINGS
-    #[cfg(test)]
-    pub fn new(s: String) -> Self {
-        Memo(s)
+impl<const M: usize> From<MemoInner<M>> for Memo<M> {
+    fn from(inner: MemoInner<M>) -> Self {
+        Self { inner: vec![inner] }
     }
 }
 
-/// Data inside a transaction
-#[derive(Clone, Debug, Eq, PartialEq)]
-pub struct Data(ByteVec);
+impl<const M: usize> TryFrom<Either<String, ByteVec>> for Memo<M> {
+    type Error = ManyError;
 
-impl<C> Encode<C> for Data {
-    fn encode<W: encode::Write>(
-        &self,
-        e: &mut Encoder<W>,
-        _: &mut C,
-    ) -> Result<(), encode::Error<W::Error>> {
-        e.bytes(self.0.as_ref()).map(|_| ())
+    fn try_from(s: Either<String, ByteVec>) -> Result<Self, Self::Error> {
+        Ok(Self {
+            inner: vec![match s {
+                Either::Left(str) => MemoInner::String(str),
+                Either::Right(bstr) => MemoInner::ByteString(bstr),
+            }],
+        })
     }
 }
 
-impl<C> Decode<'_, C> for Data {
-    fn decode(d: &mut Decoder<'_>, _: &mut C) -> Result<Self, decode::Error> {
-        let b = d.bytes()?;
-        if b.len() > MULTISIG_MEMO_DATA_MAX_SIZE {
-            return Err(minicbor::decode::Error::message("Data size over limit"));
-        }
-        Ok(Data(b.to_vec().into()))
+impl<const M: usize> TryFrom<String> for Memo<M> {
+    type Error = ManyError;
+    fn try_from(s: String) -> Result<Self, Self::Error> {
+        Ok(Self::from(MemoInner::<M>::try_from(s)?))
+    }
+}
+
+impl<const M: usize> TryFrom<ByteVec> for Memo<M> {
+    type Error = ManyError;
+    fn try_from(b: ByteVec) -> Result<Self, Self::Error> {
+        Ok(Self::from(MemoInner::<M>::try_from(b)?))
     }
 }
 
-impl TryFrom<Vec<u8>> for Data {
-    type Error = String;
+impl<const M: usize> TryFrom<Vec<u8>> for Memo<M> {
+    type Error = ManyError;
     fn try_from(b: Vec<u8>) -> Result<Self, Self::Error> {
-        if b.len() > MULTISIG_MEMO_DATA_MAX_SIZE {
-            return Err(format!("Data size over limit {}", b.len()));
-        }
-        Ok(Data(b.into()))
+        Ok(Self::from(MemoInner::<M>::try_from(b)?))
     }
 }
 
-impl Data {
-    // USED ONLY IN TESTS! NEVER REMOVE THIS GUARD! MEMOS SHOULDN'T BE PRODUCED FROM LARGE STRINGS
-    #[cfg(test)]
-    fn new(b: ByteVec) -> Self {
-        Data(b)
+impl<C, const M: usize> Encode<C> for Memo<M> {
+    fn encode<W: encode::Write>(
+        &self,
+        e: &mut Encoder<W>,
+        _: &mut C,
+    ) -> Result<(), encode::Error<W::Error>> {
+        e.encode(&self.inner).map(|_| ())
     }
+}
 
-    pub fn as_bytes(&self) -> &[u8] {
-        self.0.as_slice()
+impl<'b, C, const M: usize> Decode<'b, C> for Memo<M> {
+    fn decode(d: &mut Decoder<'b>, ctx: &mut C) -> Result<Self, decode::Error> {
```

Comment:
It's possible to decode an empty `Memo` from here, which, as previously stated, is a wrong behavior according to the specification.

Reply:
Done and added a test.